### PR TITLE
Pass along mode to BrushEvent

### PR DESCRIPTION
**Pull request by woutervh-**
Hello.

In my project I need to know the `mode` of the brush in the event handler.
This is because depending on which axis (x/y) the user is resizing the selection in, or whether the user is moving the selection without resizing (dragging), I need to change my logic on the snapping/clipping of the brush selection.

In this PR is my attempt at including the `mode` in `BrushEvent`.
I've tested it in my local project and it works fine.
Not sure if this is how you would solve it, so feel free to edit it.

### CODE REVIEW

**Comment by Fil** [comment]
We are almost ready to release d3-brush@2 with a modified API, so you will probably have to rebase it on https://github.com/d3/d3-brush/pull/71 (or wait a few days for it to be merged).

Can you share a (real world) example of how this is useful?

**Comment by woutervh-** [comment]
Hi @Fil ,

Sure, here is our real world scenario:

We allow our users to draw 2D boxes in the plot, to indicate areas of interest.
We have a requirement that these boxes cannot overlap.
After the brush event happens, we want to `.move` the selection so that we can ensure the selection does not overlap with existing boxes.

For example: if the user is using the "north" handle to vertically expand the selection, and at some point reaches an existing box, then we ensure that the "north" handle does not move past the border of the existing box.

The way that the selection needs to be `.move`'ed depends heavily on the `mode` of the brush.
This is because it's OK to alter the "north" position of the selection when the user is attempting to move it, but when the user is dragging the entire selection, it would not be good to only alter the "north" position. Instead, the size of the box needs to be conserved and the center needs to be constrained so that the edges don't cause any overlap.

Hope this makes sense, and explains why we need information about the mode.

**Comment by Fil** [comment]
v2 is now in master. It emits [a specific brushEvent](https://github.com/d3/d3-brush/blob/master/src/event.js) on each event.

That event is structured like so:
~~~js
{
  type: "brush", // or "start", "end"…
  sourceEvent: MouseEvent {…},
  target: …,
  selection: [[x0, y0], [x1, y1]]
}
~~~

I think that this is where we would now add `mode: mode.name`?

**Comment by woutervh-** [comment]
Thanks @Fil 

- Added `mode` to README
- Resolved merge conflicts from branch `two`

File: src/event.js [hunk 1]
```diff
@@ -2,13 +2,15 @@ export default function BrushEvent(type, {
...
   selection,
+  mode,
   dispatch
 }) {
   Object.defineProperties(this, {
     type: {value: type, enumerable: true, configurable: true},
     sourceEvent: {value: sourceEvent, enumerable: true, configurable: true},
     target: {value: target, enumerable: true, configurable: true},
     selection: {value: selection, enumerable: true, configurable: true},
+    mode: {value: mode, enenumerable: true, configurable: true},
```
Comment (Fil) [inline src/event.js:13]:
```suggestion
    mode: {value: mode, enumerable: true, configurable: true},
```

File: src/brush.js [hunk 2]
```diff
@@ -407,7 +408,7 @@ function brush(dim) {
     }
 
     redraw.call(that);
-    emit.start();
+    emit.start(undefined, mode.name);
```
Comment (Fil) [inline src/brush.js:411]:
```suggestion
    emit.start(event, mode.name);
```

I was really surprised by the undefined event here. Turned out this was a bug in my branch!

File: src/brush.js [hunk 3]
```diff
@@ -323,6 +323,7 @@ function brush(dim) {
           sourceEvent: event,
           target: brush,
           selection: dim.output(this.state.selection),
+          mode: mode,
```
Comment (Fil) [inline src/brush.js:326]:
```suggestion
          mode,
```

ES6 :)

**Review by Fil — CHANGES_REQUESTED** [review]
Works well 👍

**Comment by woutervh-** [comment]
Good catch in the review, how did I miss this?? `enenumerable: true`
All suggestions applied.

**Review by Fil — APPROVED** [review]

**Comment by woutervh-** [comment]
FYI: here's my attempt at a solution to stop the brush from intersecting other rectangles.

Some improvements can be made, but I will need more information from the brush's state to achieve it.
Maybe in a future issue/PR I will address these issues, but for now this is good enough for me:

https://observablehq.com/@woutervh-/brush-with-collision-solution-continued

**Pull request merged**